src/core/slice.c

```diff
@@ -385,6 +389,58 @@ static int slice_freezer_action(Unit *s, FreezerAction action) {
         return unit_cgroup_freezer_action(s, action);
 }
 
+unsigned slice_get_currently_active(Slice *slice, Unit *ignore, bool with_pending) {
+        Unit *u = UNIT(slice);
+        assert(slice);
+
+        /* If 'ignore' is non-NULL and a unit contained in this slice (or any below) we'll ignore it when
+         * counting. */
+
+        unsigned n = 0;
+        Unit *member;
+        UNIT_FOREACH_DEPENDENCY(member, u, UNIT_ATOM_SLICE_OF) {
+                if (member == ignore)
+                        continue;
+
+                if (!UNIT_IS_INACTIVE_OR_FAILED(unit_active_state(member)) ||
+                    (with_pending && member->job && IN_SET(member->job->type, JOB_START, JOB_RESTART, JOB_RELOAD)))
+                        n++;
+
+                if (member->type == UNIT_SLICE)
+                        n += slice_get_currently_active(SLICE(member), ignore, with_pending);
```

Comment:
I wouldn't make this hierarchical otherwise it'd be difficult to come up with a good values for the concurrency limit. I understand concurrency to be a slice-specific resource, not a global one.

Reply:
Maybe units can have ConcurrencyWeight= in the future. That'd allow selected units to bypass any parental limit (set to 0) and with some special value for slice units represent sum of children. But only later if it turns out useful.

Reply:
well, regarding difficulty: if people make very complex slice hierarchies it's really their own fault. these are advanced features, hence i think we can expect some higher level of udnerstanding from the user, and that they are careful with setting up a simple enough slice hierarchy that they can still make sense of everything. or alternatively they can on their own decide they never want to use the new props recursively, and hence only configure the limits on the "leave slices".

I do not understand what ConcurrencyWeight= is supposed to be? i don't get it?

Reply:
the thing is by making it hierarchial people can always use it flat too, i.e. only in the "leave" slices, and simply not bother with the inner slices for it. But the opposite wouldn't work.

Reply:
I still have troubles assigning this concurrency directive to any particular resource (there might be sleeping services or straightforward RemainAfterExit=yes, slices alone (active but empty) are (almost) nothing tangible). (BTW, if this were exclusively about CPU, there might be future [scheduler support](https://fosdem.org/2025/schedule/event/fosdem-2025-6283-a-new-cgroup-cpu-max-concurrency-controller-interface-file/).)

One more thing occurs to me -- parent slices are RequiredBy= and Before= children slices, so with the hierarchical model one _must_ account for the whole subtree otherwise nothing starts.

Anyway, since this is merged, I'll post a [fixup](https://github.com/systemd/systemd/pull/37240) and will hope for competent users.